scorometer/main.py

```diff
@@ -19,202 +27,376 @@
+            to_play.done = True
+            self.sendScore(message.id, timingScore, timingInformation)
+
+    def handleNoteOnPractice(self, message: NoteOnMessage):
+        _key = message.note
+        timestamp = message.time
+        is_down = any(x[0] == _key for x in self.keys_down)
+        if not is_down:
+            self.keys_down.append((_key, timestamp))
+            logging.debug({"note": _key})
+
+    def handleNoteOffPractice(self, message: NoteOffMessage):
+        _key = message.note
+        timestamp = message.time
+        # is_down = any(x[0] == _key for x in self.keys_down)
+        down_since = next(since for (h_key, since) in self.keys_down if h_key == _key)
+        self.keys_down.remove((_key, down_since))
+        key = Key(_key, down_since, (timestamp - down_since))
+        keys_to_play = next(
+            (i for i in self.practice_partition if any(x.done is not True for x in i)),
+            None,
+        )
+        if keys_to_play is None:
+            logging.info("Key sent but there is no keys to play")
+            self.score -= 50
+            return
+        to_play = next(
+            (i for i in keys_to_play if i.key == key.key and i.done is not True), None
+        )
+        if to_play is None:
+            self.score -= 50
+            logging.info("Invalid key.")
+        else:
+            timingScore, _ = self.getTiming(key, to_play)
+            self.score += (
+                100
+                if timingScore == "perfect"
+                else 75
+                if timingScore == "great"
+                else 50
+            )
+            to_play.done = True
+            self.sendScore(message.id, timingScore, "practice")
+
+    def getTiming(self, key: Key, to_play: Key):
+        return self.getTimingScore(key, to_play), self.getTimingInfo(key, to_play)
+
+    def getTimingScore(self, key: Key, to_play: Key):
+        tempo_percent = abs((key.duration / to_play.duration) - 1)
+        if tempo_percent < 0.3:
+            timingScore = "perfect"
+        elif tempo_percent < 0.5:
+            timingScore = "great"
+        else:
+            timingScore = "good"
+        return timingScore
+
+    def getTimingInfo(self, key: Key, to_play: Key):
+        return (
+            "perfect"
+            if abs(key.start - to_play.start) < 200
+            else "fast"
+            if key.start < to_play.start
+            else "late"
+        )
+
+    # is it in the 500 ms range
+    def is_timing_close(self, key: Key, i: Key):
+        return abs(i.start - key.start) < 500
+
+    def handleMessage(
+        self,
+        message: StartMessage
+        | EndMessage
+        | NoteOnMessage
+        | NoteOffMessage
+        | PauseMessage
+        | InvalidMessage,
+        line: str,
+    ):
+        match message:
+            case InvalidMessage(error):
+                logging.warning(f"Invalid message {line} with error: {error}")
+                send({"error": "Invalid message sent"})
```

Comment:
The error message should be explicit. Since we don't have confidential data in the scorometer you could just send the error as done above in the logging.warning call.